File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector <int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int rb = 0, re = m-1;
        int cb = 0, ce = n-1;
        vector <int> res;
        if (matrix.empty()) {
            return res;
        }
        while(rb<=re && cb<=ce)
        {
            for(int i=cb;i<=ce;i++)
                  res.push_back(matrix[rb][i]);
            rb++;
            for(int i=rb;i<=re;i++)
                  res.push_back(matrix[i][ce]);
            ce--;
            if(rb<=re){
            for(int i=ce;i>=cb;i--)
                  res.push_back(matrix[re][i]);
            re--;
               }
        if(cb<=ce)
        {
            for(int i=re;i>=rb;i--)
                 res.push_back(matrix[i][cb]);
            cb++;     
        }
                  
        }
       
        return res;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int n = matrix.length;
        List<Integer> res = new ArrayList<>();
        if(n==0) return res;
        int rb = 0;
        int re = matrix.length-1;
        int cb = 0;
        int ce = matrix[0].length-1;
        while(rb<=re && cb<=ce)
        {
            for(int i=cb;i<=ce;i++)
                res.add(matrix[rb][i]);
            rb++;
            for(int i=rb;i<=re;i++)
                res.add(matrix[i][ce]);
            ce--;
            if(rb<=re)
            {
                for(int i=ce;i>=cb;i--)
                    res.add(matrix[re][i]);
            }
            re--;
            if(cb<=ce)
            {
                for(int i=re;i>=rb;i--)
                    res.add(matrix[i][cb]);
            } 
            cb++;               
        }  
        return res;
    }
}